src/Engine/Shaders/ParticleShader.cpp

```diff
@@ -0,0 +1,273 @@
+    GLSL::checkError();
+
+    glGenTextures(1, &cellSpecularScalesTexture);
+    glActiveTexture(GL_TEXTURE0 + cellSpecularScalesTexture);
+    glBindTexture(GL_TEXTURE_1D, cellSpecularScalesTexture);
+    glTexStorage1D(GL_TEXTURE_1D, 1, GL_R32F, 16 * 4);
+    GLSL::checkError();
+
+    return true;
+}
+
+void ParticleShader::render(const CameraComponent * camera, const Vector<Component *> & components) {
+    if (!camera) {
+        return;
+    }
+
+    if (showWireFrame) {
+        glPolygonMode(GL_FRONT_AND_BACK, GL_LINE);
+    }
+
+    /* Bind uniforms */
+    loadMat4(getUniform("P"), camera->getProj());
+    loadMat4(getUniform("V"), camera->getView());
+    loadVec3(getUniform("lightDir"), *lightDir);
+    loadVec3(getUniform("camPos"), camera->gameObject().getSpatial()->position());
+
+    /* Toon shading */
+    loadFloat(getUniform("silAngle"), silAngle);
+    loadFloat(getUniform("numCells"), (float)numCells);
+    loadInt(getUniform("cellIntensities"), cellIntensitiesTexture);
+    loadInt(getUniform("cellDiffuseScales"), cellDiffuseScalesTexture);
+    loadInt(getUniform("cellSpecularScales"), cellSpecularScalesTexture);
+
+    // TODO : move cell intensities and scales to material and initialize it during json pass
+    // TODO : only upload this data once when the material is loaded in
+    glActiveTexture(GL_TEXTURE0 + cellIntensitiesTexture);
+    glBindTexture(GL_TEXTURE_1D, cellIntensitiesTexture);
+    glTexSubImage1D(GL_TEXTURE_1D, 0, 0, int(cellIntensities.size()), GL_RED, GL_FLOAT, cellIntensities.data());
+    glActiveTexture(GL_TEXTURE0 + cellDiffuseScalesTexture);
+    glBindTexture(GL_TEXTURE_1D, cellDiffuseScalesTexture);
+    glTexSubImage1D(GL_TEXTURE_1D, 0, 0, int(cellDiffuseScales.size()), GL_RED, GL_FLOAT, cellDiffuseScales.data());
+    glActiveTexture(GL_TEXTURE0 + cellSpecularScalesTexture);
+    glBindTexture(GL_TEXTURE_1D, cellSpecularScalesTexture);
+    glTexSubImage1D(GL_TEXTURE_1D, 0, 0, int(cellSpecularScales.size()), GL_RED, GL_FLOAT, cellSpecularScales.data());
+
+
+    for (ParticleComponent * pc : ParticleSystem::s_particleComponents) {
+        if (pc->getParticlePositions()->size() == 0) {
+            break;
+        }
+
+        // The reason there is a plus one is that there has to always be at least one element in RandomMs even if unused.
+        if (pc->RandomMs().size() > pc->MAX_ORIENTATIONS + 1) {
+            break;
+        }
+
+        /* Toon shading */
+        if (showToon) {
+            loadBool(getUniform("isToon"), true);
+        }
+        else {
+            loadBool(getUniform("isToon"), false);
+        }
+
+        bool r = (int)pc->RandomMs().size() != 1;
+        loadBool(getUniform("randomOrientation"), r);
+
+        /* Model matrix */
+        
+        /* Load Random Matrix Array */
+        loadMat4(getUniform("M"), pc->ModelMatrix());
+        loadMultiMat4(getUniform("randomMs"), &(pc->RandomMs()[0]), (int)pc->RandomMs().size());
```

Comment:
Lots of data being uploaded per particle per frame. Can this be fixed somehow?

Reply:
It's an extra int and a vec3 per particle. The uniforms are uploaded once for every instance, right?

Reply:
It's an extra int and a vec3 per particle. The uniforms are uploaded once for all instances, right?

Reply:
What extra int and vec3? What I see is an array of mat4s and below an array of mat3s. Are these matrix arrays unique to individual particle components? 
Regardless, like I mentioned above this is potentially 2mb being uploaded per particle component. I think there's a more efficient way of going about this.

Reply:
I think you mean 2kb. If a mat4 + mat3 is 25 floats and float is 4 bytes longs, that's 25 * 4 bytes which is 100 and 21 orientations is 2100 bytes or 2.1kb.

Reply:
Oh dam you right. Again I don't how this all plays with instancing - it might be worth bringing this up to Ian

Reply:
these random matrices should be generated once for the particle system in general and uploaded once in init.